src/filesystem.cc

```diff
@@ -1898,8 +1970,115 @@ S3FileSystem::DeleteDirectory(const std::string& path)
 
 #endif  // TRITON_ENABLE_S3
 
+
+class FileSystemManager {
+ public:
+  // create a FileSystem from the given path
+  // the class instance is to keep track and auto delete unused FileSystem
+  // the FileSystem will live for the duration of this class instance
+  // use a different class instance for different paths
+  Status GetFileSystem(const std::string& path, FileSystem** file_system);
+  // create a FileSystem from the given type
+  Status GetFileSystem(FileSystemType type, FileSystem** file_system);
+
+  FileSystemManager();
+  ~FileSystemManager();
+
+ private:
+  Status ReturnErrorOrReload(
+      const Status& load_status, const Status& error_status,
+      const std::string& path, FileSystem** file_system);
+
+  static Status LoadCredentialFromFile();
+  static Status LoadCredential();
+
+  template <class FileSystemCredential>
+  static void SortCredential(
+      std::vector<std::pair<std::string, FileSystemCredential>>& credential);
+  template <class FileSystemCredential>
+  static Status GetLongestMatchingCredential(
+      const std::vector<std::pair<std::string, FileSystemCredential>>&
+          credentials,
+      const std::string& file_system_prefix, const std::string& path,
+      FileSystemCredential& credential);
+
+  static std::recursive_mutex
```

Comment:
Can this be done with a `std::mutex` instead?

Reply:
`LoadCredential` will call `LoadCredentialFromFile` which both hold the same mutex. It might be easier to use `recursive_mutex` to avoid the same thread from locking on the same mutex twice. 

In terms of making the locking mechanism simple, I think it depends on whether we want to protect the `bool is_credential_cached` variable. Because if credential is not cached, then it is loaded from disk. By protecting this check, we prevents a possible duplicate load from disk from concurrent calls into `LoadCredential` or `LoadCredentialFromDisk`, and force concurrent calls to use cached results. On the other hand, if credential is cached, it terminates right away and release the lock, which should not be any significant reduction in parallism. I am not not sure if there is an easier way to protect this corner case?

Reply:
It looks like LoadCredentialFromFile is only called directly [here](https://github.com/triton-inference-server/core/blob/84429c6f2cca2181e9a92e3240f54a65bc662b62/src/filesystem.cc#L2248), so maybe you could tweak the logic so `LoadCredential` is the only function ever called, and it calls `LoadCredentialFromFile` internally with the lock already held?

Or something along these lines, there can be a documented expectation that the lock must be held when calling this method, etc.

Reply:
I see what you means, grouped them into a single `LoadCredential`.